Data/HashMap/Internal.hs

```diff
@@ -210,10 +213,40 @@ instance NFData2 Leaf where
 -- each key can map to at most one value.
 data HashMap k v
     = Empty
+    -- ^ Invariants:
+    --
+    -- * 'Empty' is not a valid sub-node. It can only appear at the root. (INV1)
     | BitmapIndexed !Bitmap !(A.Array (HashMap k v))
+    -- ^ Invariants:
+    --
+    -- * The array of a 'BitmapIndexed' node stores at least 1 and at most
+    --   @'maxChildren' - 1@ sub-nodes. (INV2)
+    -- * The number of sub-nodes is equal to the number of 1-bits in its
+    --   'Bitmap'. (INV3)
+    -- * If a 'BitmapIndexed' node has only one sub-node, this sub-node must
+    --   be a 'BitmapIndexed' or a 'Full' node. (INV4)
```

Comment:
There's another invariant that I should add: The upper `finiteBitSize @Word 0 - maxChildren - 1` bits of the `Bitmap` are all `0`.

Reply:
Correction: Only the `finiteBitSize @Word 0 - maxChildren` upper bits must be `0`. The bit at index 31 may be `1`.